src/timeline/TimelineModel.cpp

```diff
@@ -1934,11 +1934,298 @@ TimelineModel::formatPowerLevelEvent(const QString &id)
+    mtx::events::StateEvent<mtx::events::state::PowerLevels> *prevEvent = nullptr;
+    if (!event->unsigned_data.replaces_state.empty()) {
+        auto tempPrevEvent = events.get(event->unsigned_data.replaces_state, event->event_id);
+        if (tempPrevEvent) {
+            prevEvent =
+              std::get_if<mtx::events::StateEvent<mtx::events::state::PowerLevels>>(tempPrevEvent);
+        }
+    }
+
+    QString user        = QString::fromStdString(event->sender);
+    QString sender_name = utils::replaceEmoji(displayName(user));
+    // Get the rooms levels for redactions and powerlevel changes to determine "Administrator" and
+    // "Moderator" powerlevels.
+    auto administrator_power_level = event->content.state_level("m.room.power_levels");
+    auto moderator_power_level     = event->content.redact;
+    auto default_powerlevel        = event->content.users_default;
+    if (!prevEvent)
+        return "";
+
+    auto calc_affected = [&event,
+                          &prevEvent](int64_t newPowerlevelSetting) -> std::pair<QStringList, int> {
+        QStringList affected{};
+        auto numberOfAffected = 0;
+        // We do only compare to people with explicit PL. Usually others are not going to be
+        // affected either way and this is cheaper to iterate over.
+        for (auto const &[mxid, currentPowerlevel] : event->content.users) {
+            if (currentPowerlevel == newPowerlevelSetting &&
+                prevEvent->content.user_level(mxid) < newPowerlevelSetting) {
+                numberOfAffected++;
+                if (numberOfAffected <= 2) {
+                    affected.push_back(QString::fromStdString(mxid));
+                }
+            }
+        }
+        return {affected, numberOfAffected};
+    };
 
-    // TODO: power levels rendering is actually a bit complex. work on this later.
-    return tr("%1 has changed the room's permissions.").arg(name);
+    QStringList resultingMessage{};
+    // These affect only a few people. Therefor we can print who is affected.
+    if (event->content.kick != prevEvent->content.kick) {
+        auto default_message = tr("%1 has changed the room's kick powerlevel from %2 to %3.")
+                                 .arg(sender_name)
+                                 .arg(prevEvent->content.kick)
+                                 .arg(event->content.kick);
+
+        // We only calculate affected users if we change to a level above the default users PL
+        // to not accidentally have a DoS vector
+        if (event->content.kick > default_powerlevel) {
+            auto [affected, number_of_affected] = calc_affected(event->content.kick);
+
+            if (number_of_affected != 0) {
+                auto true_affected_rest = number_of_affected - affected.size();
+                if (number_of_affected > 2) {
+                    resultingMessage.append(
+                      default_message + QString::fromStdString(" ") +
```

Comment:
You should be using QStringLiteral or similar instead of doing a utf8->16 conversion :3